src/swap/defi/unizen.ts

```diff
@@ -0,0 +1,435 @@
+            tokenId: request.fromTokenId,
+            nativeAmount: request.nativeAmount
+          },
+          tokenContractAddress: fromContractAddress,
+          contractAddress: spendParams.destinationAddress
+        }
+      }
+      preTx = await request.fromWallet.makeSpend(spendInfo)
+    }
+
+    log('unizen preTx', JSON.stringify(preTx ?? {}))
+
+    const out: SwapOrder = {
+      expirationDate: spendParams.expirationDate,
+      fromNativeAmount: request.nativeAmount,
+      metadataNotes: spendParams.metadataNotes,
+      minReceiveAmount: spendParams.minReceiveAmount,
+      preTx,
+      request,
+      spendInfo,
+      swapInfo
+    }
+    return out
+  }
+
+  const out: EdgeSwapPlugin = {
+    swapInfo,
+
+    async fetchSwapQuote(req: EdgeSwapRequest): Promise<EdgeSwapQuote> {
+      const request = convertRequest(req)
+
+      let newRequest = request
+      if (request.quoteFor === 'max') {
+        if (request.fromTokenId != null) {
+          const maxAmount =
+            request.fromWallet.balanceMap.get(request.fromTokenId) ?? '0'
+          newRequest = { ...request, nativeAmount: maxAmount, quoteFor: 'from' }
+        } else {
+          newRequest = await getMaxSwappable(
+            async r => await fetchSwapQuoteInner(r),
+            request
+          )
+        }
+      }
+      const swapOrder = await fetchSwapQuoteInner(newRequest)
+      return await makeSwapPluginQuote(swapOrder)
+    }
+  }
+  return out
+}
+
+const makeSpendParams = (
+  request: EdgeSwapRequestPlugin,
+  log: EdgeLog,
+  fromChainId: number,
+  rawQuote: JsonObject
+): QuoteSpendParams => {
+  const { unizenName, makeFee } = PLUGIN_ID_UNIZEN_MAP[
+    request.fromWallet.currencyInfo.pluginId
+  ]
+
+  const evmSingleChain = asMaybe(asEvmSingleChainSwapQuote)(rawQuote)
+  if (evmSingleChain != null) {
+    const contractAddress: string | undefined =
+      UNIZEN_CONTRACTS[evmSingleChain[0].contractVersion]?.[unizenName]
+
+    if (contractAddress == null) {
+      log.warn(
+        'Contract address not found, try updating @unizen-io/unizen-contract-addresses ' +
+          evmSingleChain[0].contractVersion
+      )
+      throw new SwapCurrencyError(swapInfo, request)
+    }
+
+    return {
+      ...makeFee(evmSingleChain[0].gasPrice),
+      destinationAddress: contractAddress,
+      expirationDate: new Date(
+        evmSingleChain[0].transactionData.info.deadline * 1000
+      ),
+      memos: [
+        {
+          type: 'hex',
+          value: memoToHex(evmSingleChain[0].data)
+        }
+      ],
+      metadataNotes: `DEX Provider: ${evmSingleChain[0].protocol
+        .map(p => p.name)
+        .join(', ')}`,
+      minReceiveAmount: evmSingleChain[0].transactionData.info.amountOutMin
+    }
+  }
+
+  const crossChain = asMaybe(asCrossChainSwapQuote)(rawQuote)
+  if (crossChain != null) {
+    const memos: EdgeMemo[] = []
+    const memoValue = crossChain[0].transactionData.memo
```

Comment:
The `asCrossChainSwapQuote` is defined using `asArray`, but here we only access `crossChain[0]`. What if there are multiple tx's in the quote? Would the customer lose money, because the funds don't get fully swapped? I would feel safer either using `asTuple` to enforce a single item in the array, or somehow looping over the tx's to handle the weird case.

Reply:
It's an array of quotes and the best one is the first one. I'll change improve the cleaner name and change to asTuple to enforce it's existance